relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -358,20 +415,12 @@ pub fn sample_transaction(
 fn get_matching_rule<'a, T>(
     config: &'a SamplingConfig,
     context: &T,
-    project_id: ProjectId,
-    strategy: SamplingStrategy,
+    ty: RuleType,
 ) -> Option<&'a SamplingRule>
 where
-    T: SamplingContextProvider,
+    T: FieldValueProvider,
 {
-    let user_segment = context.user_segment().map(LowerCaseString::new);
-    let environment = context.environment().as_deref().map(LowerCaseString::new);
-    let release = context.release();
-
-    config
-        .rules
-        .iter()
-        .find(|rule| rule.matches(release, &user_segment, &environment, project_id, strategy))
+    config.rules.iter().find(|rule| matches(context, rule, ty))
```

Comment:
Instead of passing `ty` down to `matches`, you could inline this here trivially via. To me, this makes the flow easier to follow.

```suggestion
    config.rules.iter().find(|rule| rule.ty == ty && rule.condition.matches(value_provider))
```

Reply:
good point, top level function matches can go